### ISSUE TITLE
A false negative about self assignment

### ISSUE THREAD

**Issue by Belle-PL** [google/error-prone#2432]
### Description of the problem / feature request:

> A false negative about error-prone rule [**SelfAssignment**](https://errorprone.info/bugpattern/SelfAssignment)

### Feature requests: what underlying problem are you trying to solve with this feature?

> I found this false negative when wrote some test cases for project. This is an obvious case for self assignment, but error-prone can not recognize it. 

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

```java
public void testSelfAssignment() {
    long x = 10;
    x = x + 1 - 1; // Should report a warning here
}
```

Expected Outcome: Report a warning in line 3
Actual Outcome: No warning

### What version of Error Prone are you using?

> 2.7.1

**Comment by Belle-PL** [comment]
@cushon  Hi Cushon, could you please kindly help check this issue? I think it can improve error prone.

**Comment by graememorgan** [comment]
Have you seen this pattern correspond to an error in practice? I would imagine there's a plethora of similar patterns you could come up with (e.g. `x = x + "".length()`).

**Comment by cushon** [comment]
Similar to #2414, we're not trying to find all possible occurrences of a particular pattern, just the ones that are likely to be bugs and are easy to find with simple intraprocedural analysis.

e.g. the following could also be self assignment if the implementation of `f` is `long f(long x) { return x; }`, but we can't tell without interprocedural analysis:

```
void test
    long x = 10;
    x = f(x);
}

abstract void f();
```

**Comment by cushon** [comment]
@Belle-PL, are the bugs you've been filing based on issues you've found in your code base that Error Prone didn't catch? Or were they examples that you constructed while evaluating Error Prone?

I appreciate the attention to detail here, but it'd be more helpful to know if there are issues you're encountering in real-world code trying to use Error Prone.

For example, if there's code like `x = x + 1 - 1;` in your codebase that was a bug, that'd be a stronger reason to change this check (or maybe introduce a new one) than if that was a constructed example to test what cases this check can detect.
